src/galileo/utils/singleton.py

```diff
@@ -64,36 +68,50 @@ def _get_key(
             The experiment ID.
         mode: (Optional[str])
             The logger mode.
+        trace_id: (Optional[str])
+            The distributed trace ID (for async web server isolation).
+        span_id: (Optional[str])
+            The distributed parent span ID (for async web server isolation).
 
         Returns
         -------
-        Tuple[str, str, str, str]
-            A tuple key (project, log_stream, experiment_id, mode) used for caching.
+        Tuple[str, ...]
+            A tuple key used for caching. Includes trace_id and span_id for proper
+            isolation of concurrent requests in async web servers.
         """
         _logger.debug("current thread is %s", threading.current_thread().name)
 
         # GalileoLoggerSingleton must NOT be shared across different threads
         current_thread_name = threading.current_thread().name
         key = (current_thread_name, mode)
 
-        if project is None:
-            project = getenv("GALILEO_PROJECT", DEFAULT_PROJECT_NAME)
-        if log_stream is None:
-            log_stream = getenv("GALILEO_LOG_STREAM", DEFAULT_LOG_STREAM_NAME)
+        # Get project and log_stream with environment variable fallbacks
+        project = _get_project_or_default(project)
+        log_stream = _get_log_stream_or_default(log_stream)
+
+        # Include trace_id and span_id in key for async web server isolation
```

Comment:
I don't know if we need to say "async web server isolation" in 3 places.. maybe keep here since it has the most explanation and remove from the places above? Just a nit-picky opinion maybe, feel free to ignore